Sources/MessagingPush/PushHandling/PushEventHandlerProxy.swift

```diff
@@ -39,8 +39,18 @@ class PushEventHandlerProxyImpl: PushEventHandlerProxy {
             return
         }
 
+        let numberOfAsyncOperationsExpected = nestedDelegates.count
+        var numberOfAsyncOperationsCompleted = 0
+
         nestedDelegates.forEach { _, delegate in
-            delegate.onPushAction(pushAction, completionHandler: completionHandler)
+            delegate.onPushAction(pushAction, completionHandler: {
+                numberOfAsyncOperationsCompleted += 1
+
+                // Do not call the function's completionHandler until *all* delegates have had their completion handler called.
+                if numberOfAsyncOperationsExpected == numberOfAsyncOperationsCompleted {
+                    completionHandler()
+                }
+            })
```

Comment:
a couple of things, 
- using DispatchGroup over variable tracking could help with synchronization and thread safety
- can we include timeout to ensure, `completionHandler()` is called in case it's not called by the user? something like this? 



```
func onPushAction(_ pushAction: PushNotificationAction, completionHandler: @escaping () -> Void) {
    let dispatchGroup = DispatchGroup()
    let timeoutInterval = 10.0 // Timeout after 10 seconds?

    for delegate in nestedDelegates.values {
        dispatchGroup.enter()
        delegate.onPushAction(pushAction) {
            dispatchGroup.leave()
        }
    }

    // Completion block to execute once all handlers have completed.
    let completionBlock = {
        // All delegates have processed the push action, or a timeout occurred.
        // Proceed with CIO SDK-specific logic.
        completionHandler()
    }

    // Set up the completion notification.
    dispatchGroup.notify(queue: .main, execute: completionBlock)

    // Set up a timeout mechanism.
    DispatchQueue.main.asyncAfter(deadline: .now() + timeoutInterval) {
        completionBlock()
    }
}
```

Reply:
This suggestion is worrisome. If my understand is correct, then if a callback in any delegate isn't called the `dispatchGroup.leave()` won't be called.

Then we pretty much have a queue handing forever, yes it is non-blocking, but it still a resource that's created and for every push and remained there.

Which seems reasonable to say that we are trying to synchronise with async code (the `delegate .onPushAction`) that we don't own. While this can seem like a trivial problem, it is not and we need to be careful how we proceed.

Specially the fact that we needed to introduce timeout, if you think about it, this is what the OS watchdog does, and this is no trivial task either.

Let's talk more about this in the comms window to make sure at least we are on the same page of what's the problem we are solving

Reply:
Conversation happening in [the other PR](https://github.com/customerio/customerio-ios/pull/584#discussion_r1509523944).